src/System.Linq/src/System/Linq/Intersect.cs

```diff
@@ -19,8 +19,28 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
             {
                 throw Error.ArgumentNull(nameof(second));
             }
+           
+            IEnumerable<TSource> big = null;
+            IEnumerable<TSource> small = null;
 
-            return IntersectIterator(first, second, null);
+            if (first.Count() > second.Count())
+            {
+                big = first.Distinct();
+                small = second.Distinct();
+            }
+            else
+            {
+                big = second.Distinct();
+                small = first.Distinct();
+            }
+
+            foreach (TSource element in big)
+            {
+                if (small.Contains(element))
```

Comment:
I mean that `first` and/or `second` could be yielding from the results of a database or web service, and this change means instead of doing so once, it may do so multiple times.